### PR TITLE
Move overrides() and leastAsSpecialized() to funcsem.d

### PR DESCRIPTION
**Pull request by muscar**
Part of the effort to extract semantic analysis code from the AST code.

### CODE REVIEW

File: compiler/src/dmd/funcsem.d [hunk 1]
```diff
@@ -1931,6 +1931,112 @@ FuncDeclaration overloadExactMatch(FuncDeclaration thisfd, Type t)
...
+ * This is very similar to TemplateDeclaration::leastAsSpecialized().
+ * Params:
+ *  f = first function
+ *  g = second function
+ *  names = names of parameters
+ * Returns:
+ *      match   'this' is at least as specialized as g
+ *      0       g is more specialized than 'this'
+ */
+MATCH leastAsSpecialized(FuncDeclaration f, FuncDeclaration g, Identifiers* names)
```
Comment (ibuclaw) [inline compiler/src/dmd/funcsem.d:1966]: You'll need to add this to the dmd namespace in [cxxfrontend.d](https://github.com/dlang/dmd/blob/e4eae5990c10dfa030a7cb126cb3099791aaeb73/compiler/src/dmd/cxxfrontend.d) for C++ backends.

Then [move](https://github.com/dlang/dmd/blob/e4eae5990c10dfa030a7cb126cb3099791aaeb73/compiler/src/dmd/declaration.h#L709) the declaration [here](https://github.com/dlang/dmd/blob/e4eae5990c10dfa030a7cb126cb3099791aaeb73/compiler/src/dmd/declaration.h#L36).

File: compiler/src/dmd/funcsem.d [hunk 2]
```diff
@@ -1931,6 +1931,112 @@ FuncDeclaration overloadExactMatch(FuncDeclaration thisfd, Type t)
     return fd;
 }
 
+/****************************************************
+ * Determine if fd1 overrides fd2.
+ * Return !=0 if it does.
+ */
+extern (D) int overrides(FuncDeclaration fd1, FuncDeclaration fd2)
```
Comment (ibuclaw) [inline compiler/src/dmd/funcsem.d:1938]: `extern(D)` would be redundant here.

File: compiler/src/dmd/funcsem.d (hunk 1 above)
Comment (muscar) [reply compiler/src/dmd/funcsem.d:1966]: Cheers, done.

File: compiler/src/dmd/funcsem.d (hunk 2 above)
Comment (muscar) [reply compiler/src/dmd/funcsem.d]: Done.

**Review by ibuclaw — COMMENTED** [review]
Tests are failing as they need updating too.

```
/home/runner/work/dmd/dmd/dmd/compiler/src/tests/cxxfrontend.cc:1343:38: error: ‘leastAsSpecialized’ is not a member of ‘FuncDeclaration’
```

**Review by thewilsonator — APPROVED** [review]

**Comment by muscar** [comment]
Thanks for the reviews @ibuclaw @thewilsonator

**Pull request merged**